src/hooks/useLocalStorage.ts

```diff
@@ -1,32 +1,71 @@
 import { useState } from "react";
 
+type LocalStorageValue = number | string | boolean | {};
+export type LocalStorageInitCb<T extends LocalStorageValue> = (
+  initial: unknown
+) => T;
+
 export type SetLocalStorageValue<T> = T | ((prevValue: T) => T);
 export type SetLocalStorage<T> = (value: SetLocalStorageValue<T>) => void;
-export default function useLocalStorage<T>(
+export type UseLocalStorage<T> = [T, SetLocalStorage<T>];
+export default function useLocalStorage<T = LocalStorageValue>(
   key: string,
-  initialValue: T
-): [T, SetLocalStorage<T>] {
+  initialValue: T | LocalStorageInitCb<T>
+): UseLocalStorage<T> {
+  const saveToLocalStorage = (valueToStore: T) => {
+    try {
+      if (typeof valueToStore === "string") {
+        localStorage.setItem(key, valueToStore);
+      } else {
+        localStorage.setItem(key, JSON.stringify(valueToStore));
+      }
+    } catch {
+      console.warn(`Could not save ${key} to localStorage`);
+    }
+  };
+
+  function getValue<T extends LocalStorageValue>(
+    value: T,
+    initOrCb: T | LocalStorageInitCb<T>
+  ): T {
+    if (typeof initOrCb === "function") {
+      const newValue = initOrCb(value);
+      saveToLocalStorage(newValue);
+      return newValue;
+    }
+
+    return value ?? initOrCb;
+  }
+
   const [storedValue, setStoredValue] = useState<T>(() => {
-    let result: T;
+    let result: T | null;
+    const item = localStorage.getItem(key);
+
+    if (item === null) {
+      return getValue(null, initialValue);
+    }
+
     try {
-      const item = window.localStorage.getItem(key);
-      result = item ? JSON.parse(item) : initialValue;
+      const parsed = JSON.parse(item);
```

Comment:
```suggestion
      result = JSON.parse(item);
```

Reply:
I know it can be shortened to this, but I found declaring transitive variable more readable